### PR TITLE
[WIP] `init` endpoint of ui-server

### PR DESCRIPTION
**Pull request by LakshSingla**
**What kind of change does this PR introduce?**
Feature

**Did you add tests for your changes?**
No

**If relevant, did you update the documentation?**
No

**Summary**
Created the route for `POST: init` for default scaffold. Does not work as expected due to issues in `stdin` of the child process spawned.

#649 

**Does this PR introduce a breaking change?**
No

**Other information**
None

### CODE REVIEW

File: packages/ui-server/app.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ const path = require("path");
 const cookieParser = require("cookie-parser");
 const logger = require("morgan");
 const indexRouter = require("./routes/index");
+const apiRouter = require("./routes/api");
```
Comment (rishabh3112) [inline packages/ui-server/app.js]: Redundant import
```suggestion

```

File: packages/ui-server/routes/index.js [hunk 2]
```diff
@@ -3,7 +3,9 @@ const router = express.Router();
 
 /* GET home page. */
 router.get("/", function(req, res, next) {
-	res.render("index", { title: "Express" });
+	res.json({ title: "Express" });
 });
 
+router.use(require('./api'));
```
Comment (rishabh3112) [inline packages/ui-server/routes/index.js]: use `/api` as the base route for the server calls.
We would be using `/api/init` for json calls to render `/init` and same for other endpoints.
```suggestion
router.use('/api',require('./api'));
```

**Comment by rishabh3112** [comment]
Sign CLA too

**Comment by LakshSingla** [comment]
@rishabh3112 Thanks for the suggestions. I have made the required changes. But I am still stuck on how to call the actual `init` process. I have used `execa` to create a child process and passed in the options to its `stdin` (which I receive in the request). But I am facing an issue because of the way I am passing the `stdin` due to which all output is dumped into the child process at once and it cannot read it for subsequent queries.
webpack-init-1

**Comment by rishabh3112** [comment]
I think we should refactor and change interface of `webpack-scaffold` all together so as to avoid running child process. A configuration generating file which will take answers to questions as json input and generates file.

This way debugging in future would be better.Let have others viewpoint. 
cc @evenstensberg @ematipico

**Comment by LakshSingla** [comment]
@rishabh3112 I was thinking the same thing, to provide an alternative way of programmatically passing in the parameters for `init`.

**Comment by LakshSingla** [comment]
@rishabh3112 Should I start with the refactoring? Also what all needs to be refactored? I would begin with the `webpack init` command located at `@webpack-cli/init`.

**Comment by ematipico** [comment]
Yes, inquirier logic and generator logic have to be separated in order to achieve this PR.

The server will just pass an object to the generator, which will create the webpack file for us. The object will contain the same set of answers.

Because of the fact that we are using typescript, I think there could be an interface that design the object answers that we can pass to the generator. If there isn't, we should create it.

I think to make things easier, using typescript for the server might be a better fit because of the typings...

**Comment by LakshSingla** [comment]
@ematipico Should I open a seperate issue regarding this issue? Also, given that the generator logic is in place `@webpack-cli/packages/generators/index-generator.ts`, I propose that we move the inquirer logic to another file.

**Comment by rishabh3112** [comment]
I think better would be that every generator exports:
1. `cli`: would be **default** class
2. `ui`: would be processing Object.

and also a function that would return a multi-dimensional array having **questions**.

Then for passing Object we can:
```js
import { ui } from "@webpack-cli/init"
...
let input: Object; // or we may have a interface here instead of object
// taking input from server
ui(input);
...
```
Review approch @ematipico I would do necessary updations then. Also, @evenstensberg didn't gave review for design as he said to me he will be. If you have suggestions you may let me know. After that I would design other pages as well.

**Comment by LakshSingla** [comment]
@rishabh3112 Thanks for help with the PR! As soon as you are done with the necessary refactoring, lemme know so that I can make the required changes.

**Comment by LakshSingla** [comment]
@rishabh3112 Any updates in refactoring ??

**Comment by evenstensberg** [comment]
No worries

File: packages/ui-server/bin/www/index.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+const path = require('path');
+const app = require('../../app.js');
+
+const port = process.env.PORT || 8000;
+
+console.log(port);
```
Comment (evenstensberg) [inline packages/ui-server/bin/www/index.js]: remove log here

File: packages/ui-server/bin/www/index.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+app(path.join('..', '..', 'static')).listen(port, () => {
```
Comment (evenstensberg) [inline packages/ui-server/bin/www/index.js]: extract to var

File: packages/ui-server/bin/www/index.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+	console.log(`server started on 127.0.0.1:${port}`);
```
Comment (evenstensberg) [inline packages/ui-server/bin/www/index.js]: use webpack logger

Comment (LakshSingla) [reply packages/ui-server/bin/www/index.js]: Made the necessary changes

**Comment by rishabh3112** [comment]
#700

**Comment by fokusferit** [comment]
hey @rishabh3112 any update on this PR ?

**Comment by fokusferit** [comment]
Just to understand this correctly, this one can be close as @rishabh3112 continue the work at #756 right?

**Comment by wizardofhogwarts** [comment]
Yes.

**Comment by evenstensberg** [comment]
Could you rebase this?

**Comment by evenstensberg** [comment]
As per meetings, we're decided that UI should be third party

**Comment by evenstensberg** [comment]
Thank you so much for taking the time to create this Pull Request, you've helped shape the way we're going with webpack(cli) ❤️

**Pull request closed without merging**